x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticMultiMatchQueryRewriteInterceptor.java

```diff
@@ -0,0 +1,487 @@
+        disMaxQuery.queryName(originalQuery.queryName());
+        return disMaxQuery;
+    }
+
+    /**
+     * Builds a most_fields query for pure semantic fields using BoolQueryBuilder.
+     */
+    private QueryBuilder buildMostFieldsSemanticQuery(
+        MultiMatchQueryBuilder originalQuery,
+        Map<String, Float> fieldsBoosts,
+        Set<String> inferenceFields,
+        String queryValue
+    ) {
+        BoolQueryBuilder boolQuery = new BoolQueryBuilder();
+        for (String fieldName : inferenceFields) {
+            boolQuery.should(createSemanticQuery(fieldName, queryValue, fieldsBoosts, false));
+        }
+        boolQuery.minimumShouldMatch("1");
+        boolQuery.boost(originalQuery.boost());
+        boolQuery.queryName(originalQuery.queryName());
+        return boolQuery;
+    }
+
+    private QueryBuilder buildBestFieldsCombinedQuery(
+        MultiMatchQueryBuilder originalQuery,
+        Map<String, Float> fieldsBoosts,
+        MultiFieldInferenceInfo inferenceInfo,
+        String queryValue
+    ) {
+        DisMaxQueryBuilder disMaxQuery = QueryBuilders.disMaxQuery();
+
+        addInferenceQueriesPerIndex(disMaxQuery, inferenceInfo, queryValue, fieldsBoosts);
+        addNonInferenceQueriesPerIndex(disMaxQuery, inferenceInfo, originalQuery, fieldsBoosts);
+
+        // Apply tie_breaker if specified
+        if (originalQuery.tieBreaker() != null) {
+            disMaxQuery.tieBreaker(originalQuery.tieBreaker());
+        }
+        disMaxQuery.boost(originalQuery.boost());
+        disMaxQuery.queryName(originalQuery.queryName());
+        return disMaxQuery;
+    }
+
+    private QueryBuilder buildMostFieldsCombinedQuery(
+        MultiMatchQueryBuilder originalQuery,
+        Map<String, Float> fieldsBoosts,
+        MultiFieldInferenceInfo inferenceInfo,
+        String queryValue
+    ) {
+        BoolQueryBuilder boolQuery = new BoolQueryBuilder();
+
+        addInferenceQueriesPerIndex(boolQuery, inferenceInfo, queryValue, fieldsBoosts);
+        addNonInferenceQueriesPerIndex(boolQuery, inferenceInfo, originalQuery, fieldsBoosts);
+
+        boolQuery.minimumShouldMatch("1");
+        boolQuery.boost(originalQuery.boost());
+        boolQuery.queryName(originalQuery.queryName());
+        return boolQuery;
+    }
+
+    /**
+     * Detects and warns about score range mismatches when a multi_match query has at least one dense vector model (TEXT_EMBEDDING)
+     * mixed with sparse vector models (SPARSE_EMBEDDING) or non-inference fields.
+     * Dense vector models typically produce bounded scores (0-1) while sparse vector models and
+     * non-inference fields produce unbounded scores, causing score range mismatches.
+     */
+    private void detectAndWarnScoreRangeMismatch(MultiFieldInferenceInfo inferenceInfo) {
+        ModelRegistry modelRegistry = modelRegistrySupplier.get();
+        // TODO: validate if we need to check if modelRegistry is null or not
```

Comment:
I see you have a TODO here to verify this already, but I think if we don't have access to the model registry we're going to be in a bad state.